js/common/lib/backend.ts

```diff
@@ -31,6 +31,30 @@ export interface SessionHandler {
       options: InferenceSession.RunOptions): Promise<SessionHandler.ReturnType>;
 }
 
+export declare namespace TrainingSessionHandler {
+  type FeedsType = {[name: string]: OnnxValue};
+  type FetchesType = {[name: string]: OnnxValue | null};
+  // shall we change this to just error code. Right now this will return loss, but it is not
+  // as important and too much cost to read the loss just for printing purpose.
+  type ReturnType = {[name: string]: OnnxValue}|number;
+}
+
+/**
+ * Represent a handler instance of a training inference session.
+ *
+ * @internal
+ */
+export interface TrainingSessionHandler {
```

Comment:
training session handler should extends (inference) session handler, as described in design doc.

Reply:
Should a `TrainingSessionHandler` instance be able to do all InferenseSessionHendler can do?

Reply:
Ah, no, just the training backend was intended to extend the inference wasm backend. The TrainingSessionHandler should not be able to do all that the InferenceSessionHandler can do -- users should be able to start both InferenceSessions and TrainingSessions in the same piece of code, but the InferenceSession and the TrainingSession should still be two distinct things.

Reply:
No TrainingSessionHandler should be able to handle all the training related functionality like running train, eval, optimizer step, setting learning rate, reset gradients etc. 

Backend interface will be extended to include both the inference and training session handlers. 

@carzh : You can add all the training apis like runTrainStep etc behind a comment in this PR. This way you wont see any errors but at the same time reviewers will get an idea what functionality will be available via TrainingSessionHandler.

Reply:
Added the runTrainStep & parameter handling functions in comments to the training-session interface

Reply:
Although you marked it as "Added", I didn't see it added in `interface TrainingSessionHandler` defined here.

Reply:
this interface should include the training functions, like "runTrainStep()"

Reply:
Since you have some properties/methods shared, you can create the shared interface:
```ts
export interface SessionHandler {
  dispose(): Promise<void>;

  readonly inputNames: readonly string[];
  readonly outputNames: readonly string[];
}
```

and you can have corresponding interfaces for inference and training:

```ts
interface InferenceSessionHandler {
  startProfiling(): void;
  endProfiling(): void;

  run(feeds: SessionHandler.FeedsType, fetches: SessionHandler.FetchesType,
      options: InferenceSession.RunOptions): Promise<SessionHandler.ReturnType>;
}
```

```ts
interface TrainingSessionHandler {
  runTrainStep(...);
}
```
```

Reply:
this is a good suggestion but maybe this change can be bundled into a future PR because it'd require references of SessionHandler to be changed to InferenceSessionHandler in all the backends.